### ISSUE TITLE
"impossible" custom exercise

### ISSUE THREAD

**Issue by evgenyfadeev** [didoesdigital/typey-type#199]
I've come across an exercise that is quite difficult to complete, even if all the chords are entered correctly:

```
liquidity	HREU/KWEUTD
liquidity	HREU/KWEUTD
```

This issue might be related to #191 ,  and here I seem to be able to reproduce it every time.

What I see is that the first word is entered correctly, but the second one I cannot enter, because once I type in the first chord, some letters from previous chord are randomly pre-fixed.

I'm using Lapwing dictionaries on plover.

**Comment by didoesdigital** [comment]
Hmm, I think this _might_ be an awkward combination of [orthography rules](https://www.openstenoproject.org/learn-plover/lesson-7-orthography-rules-and-suffix-keys.html) and [word boundary errors](https://www.openstenoproject.org/learn-plover/appendix-word-boundaries.html) and problems caused by starting an outline with a suffix stroke.

Let's imagine you are writing in a normal text editing app (not Typey Type):

- `HREU` matches the entry `"HREU": "{^ly}",` and writes the text `ly`
- Adding `KWEUTD` means the Plover buffer has `HREU/KWEUTD`, which matches `"HREU/KWEUTD": "liquidity",` and backtracks to remove `ly ` and outputs ` liquidity` instead
- Adding `HREU` means the Plover buffer has `HREU/KWEUTD/HREU`, which does not match anything so it assumes it's the beginning of a new word like `HREU/KWEUTD HREU` and clears the buffer and starts a new word. For the new `HREU`, it again matches the entry `"HREU": "{^ly}",` so it appends `ly`. This would cause a kind of word boundary error because you wanted to start a new word but unintentionally extended the previous one. It _should_ now read `liquidityly`. However [this Plover orthography rule](https://github.com/openstenoproject/plover/blob/9b5a357f1fb57cb0a9a8596ae12cd1e84fcff6c4/plover/system/english_stenotype.py#L74-L75) kicks in and tries to replace the `y` in `ty` with `i`, resulting in `liquiditily`.
- Adding the final `KWEUTD` means the Plover buffer has `HREU/KWEUTD` and tries to undo `ly` as a suffix and instead write the new "liquidity", so it backtracks to remove `ily`  and restores `y` and adds ` liquidity`.

Now that last step to backtrack to remove `ily` and restore `y` doesn't work very well in Typey Type because Typey Type has removed the whole previous word so it just awkwardly inserts `y ` at the start of the text field.

One approach to handle this is to learn which strokes are suffix strokes and anticipate word boundary errors like this and insert a space in between e.g. `HREU/KWEUTD S-P HREU/KWEUTD`.

Another approach would be to change dictionary entries and avoid starting any regular words with a stroke that's used as a suffix stroke. For example, because `HREU` has the suffix translation `{^ly}`, a stenographer might choose to avoid having any entries for words starting with `HREU/`. So for "liquidity", you might use `HREUBG` as the first stroke because it's not a suffix stroke, and use `"HREUBG/WEUD/TEU": "liquidity"`

Does that all make sense?

**Comment by didoesdigital** [comment]
Here is a slow demo of writing `HREU/KWEUTD HREU/KWEUTD` in a text editor while showing the Plover paper tape so the individual steno keys pressed can be shown:

![liquidity](https://github.com/user-attachments/assets/0f31c413-e3a4-4d51-8158-44b451669c0b)

And here's what the stroke log shows was stroked and translated:

![stroke-log](https://github.com/user-attachments/assets/e46287b1-1f86-47b3-836d-4a90a75959f6)

**Comment by evgenyfadeev** [comment]
@didoesdigital it does make sense, thanks! 

However, I'd think that the app is behaving differently in comparison with typing normally - where the output recovers eventually. Here - unless a word termination is entered after the word succeded by a syllable coinciding with a suffix, the text in the input box won't eventually produce the expected result. In that sense I'd still think that this is an issue.

Basically - TypeyType is teaching to type differently from how one would do that in other text editors. Perhaps not a huge issue once you know the reason, but still would be better not to have it.

**Comment by didoesdigital** [comment]
Generally I think it's a good habit to avoid using suffix strokes as the start of words so if a stenographer is following that practice then I would expect the result to be that the app would behave the same as typing normally. In practice, it is very important for stenographers to know how to write without word boundary errors.

That said, it is undoubtedly confusing and weird and quite possibly frustrating too, so it's worth considering how we might make the experience better!

On a technical level, I don't know that there's much we could do in the lesson interface. There are good reasons for Typey Type accepting 1 word at a time instead of what some typing apps do like A) letting you write the entire lesson into 1 text field or B) not using a real text field but manually showing what's been typed as text. I would not be enthused about revamping the whole paradigm of the lesson typed text interface. An alternative approach would be building steno-engine kind of logic to do something like back tracking into the previously entered word, AND it would need to know what steno keys were pressed instead of only what text was typed, which would in turn probably require browser plugins or somehow… processing steno protocol inputs… 😕 Another approach might be treating a backspace in the lesson text field when it is empty as a trigger to move back to the previous word. That might unlock pandora's box with knock on effects to other logic in the app that assumes that behaviour is not possible (e.g. all the scores would be whacky). It does not fill me with joy 😅

But! What if we try to address the matter outside of the lesson interface? Something like the [TPEUBGSZ](https://didoesdigital.com/typey-type/games/TPEUBGSZ) game but focused specifically on teaching word boundary errors? It could be full of traps like "liquidity liquidity" and have a different kind of input that doesn't delete previous words. The goal might be to write the words without touching the previous word. 

I dunno, I'm just kicking around ideas here. Thoughts?
